text/0000-std-function-macro.md

```diff
@@ -0,0 +1,153 @@
+`std::module_path` and expands to an expression of type `&'static str`.
+
+# Motivation
+[motivation]: #motivation
+
+There are currently macros for getting the source file name, `std::file`, and
+the line number, `std::line`. To complete this pair, a macro to for getting
+the current function name would be useful. All three macros are useful for
+e.g. generating log output with more context about where something was
+happening. The file and line is good for being able to look up the exact
+location, however the fully qualified function name gives additional context
+without even having to look into the source code.
+
+# Detailed design
+[design]: #detailed-design
+
+This feature would have an equivalent design and implementation to
+`std::file`. It would return a fully qualified function name but the actual
+string returned is implementation defined. The string is meant to be used by
+the user to get an idea about the context of the code in e.g. debug output,
+and should be meaningful for a human, but is not meant to be analyzed by a
+program.
+
+For example the following scheme could be used:
+
+ - Plain functions use their name, e.g. `some_function` for
+
+````
+    fn some_function(arg: SomeThing) {
+    ...
+    }
+````
+
+ - Generic functions use their name, followed by the generic arguments, e.g.
+   `some_function<T, U>` for
+
+````
+    fn some_function<T, U>(arg: SomeThing) {
+    ...
+    }
+````
+
+ - Functions inside non-generic `impl` blocks for structs would use the name of the
+   struct, followed by `::` and the function name, e.g.
+   `SomeStruct::some_function` for
+
+````
+    struct SomeStruct {...}
+
+    impl SomeStruct {
+        fn some_function(&self, arg: SomeThing) {
+        ...
+    }
+````
+
+ - Functions inside generic `impl` blocks for structs would use the name of the
+   struct, followed by the generic parameters as given by the `impl` block,
+   followed by `::` and the function name, e.g.
+   `SomeStruct<i32>::some_function` for
+
+````
+    struct SomeStruct<T> {...}
+
+    impl SomeStruct<i32> {
+        fn some_function(&self, arg: SomeThing) {
+        ...
+    }
+````
+
+ - Similar for generic `impl` blocks, e.g. `SomeStruct<T>::some_function` for
+
+````
+    struct SomeStruct<T> {...}
+
+    impl<T> SomeStruct<T> {
+        fn some_function(&self, arg: SomeThing) {
+        ...
+    }
+````
+
+ - Functions inside trait `impl` blocks would use, enclosed in angle brackets,
+   the name of the type, followed by `as`, followed by the trait name, and
+   then `::` followed by the name of the function, e.g.
+   `<SomeStruct as SomeTrait>::some_function` for
+
+````
+    impl SomeTrait for SomeStruct {
+        fn some_function(&self, arg: SomeThing) {
+        ...
+    }
+````
+
+ - Closures are using the name of the surrounding function, followed by `::`
+   and the string `<closure/definition_line_number>`, where `definition_line_number` is the number
```

Comment:
I’m considerably against including information like line number into output the proposed macro. `line!()` already covers the use-case of differentiating code by line numbers in the source code quite well.

Reply:
(rust noob here, so assume I don't know much and can easily be wrong on even basic things)

**tl;dr:** We need to ask ourselves a question: do we want to uniquely identify the function in the output of `function_name!()` ? Alternatively, do we want just the function name? or, somewhere in between these two? (this `in between` variant is what we'd get by not including the line number of the function definition for certain cases: closures(which the current comment thread is addressing), or same-level [nested](https://github.com/rust-lang/rfcs/pull/1719#issuecomment-245224914) inner functions(which I'm trying to address below))

Long read:

Interesting...
@da-x brings up a good point [here](https://github.com/rust-lang/rfcs/issues/1743#issuecomment-361223593)

With regards to inner functions nested at the [same](https://github.com/rust-lang/rfcs/pull/1719#issuecomment-245224914) level, but also generally speaking:
Do we want to uniquely identify the named function ? or do we just want the function name?(knowing full well that there can be multiple functions with the same name and different signatures(eg. `some_func<T, U>` and `<SomeStruct as SomeTrait>::some_func`), or in different nesting blocks, [example](https://github.com/rust-lang/rfcs/pull/1719#issuecomment-245224914))

Perhaps a `function_name!(bool)` could handle both cases, whereby if arg is `true` then it will uniquely identify function(eg. add line number?) and if `false`, just keep the function name without any line numbers.

But then you could argue the same for the other variants that try to somewhat-uniquely identify the named function like appending `<T, U>` or prepending `<SomeStruct as SomeTrait>::` to the function name.

Maybe, the same `bool` arg could be used to affect those? So if `false` you'd strictly get ONLY the function name ie. `some_function`, but if `true` you'd get not only `<SomeStruct as SomeTrait>::some_function` but also line numbers(but only where applicable?) like `fn1/definition_line_number_of_fn1::fn2/definition_line_number_of_fn2`

Or, should there be some way to select which extra things you want in the returned function name? Like do you want the `<T, U>` in `some_function<T, U>` or the `<SomeStruct as SomeTrait>::` in `<SomeStruct as SomeTrait>::some_function` or the line number of function definition always like `some_function/76` (or whichever would be the format for showing the line number). Note here, that you can get the `line!()` number of the call, but we're talking the line number of the function definition here! which seems far more important, perhaps.

Now that I think about this: [`function_name is a macro, which expands to the fully qualified function name from which it was invoked.`](https://github.com/sdroege/rfcs/blob/standard-function-macro/text/0000-std-function-macro.md#summary) that makes me want the name to be `function_path`, and `function_name` to be something that has only the function name.
Either way, that statement seems to imply that it wants to uniquely identify the function, which means, it would want to have the function definition line number in those cases where it's a closure or inner functions nested at the [same](https://github.com/rust-lang/rfcs/pull/1719#issuecomment-245224914) level.